text/0006-shipyard-next-generation.md

```diff
@@ -0,0 +1,317 @@
+
+### Event selector
+
+A sequence has a subscription to certain events that trigger the sequence. However, an SRE not only wants to configure the event type, but also a selector on this event. This selector configures more precisely when the sequence should be triggered. For example, an SRE wants to configure that a rollback sequence should be triggered just when deployment or release failed. Therefore, this KEP proposes the approach of an event selector using  `matchLabels` and `matchExpressions`. 
+
+While the `matchLabels` allows configuring a selector on a `key` with exactly one expected `value`, the `matchExpressions` allows configuring a selector on a `key` with multiple value options. For example, the below rollback sequence has a subscription to two events. (1) For the first `hardening.deployment.finished` event a selector is configured based on the label `status:failed`. This means that the rollback sequence gets triggered when the event has the label `status:failed` attached. (2) For the second `hardening.release.finished` event a selector is configured for the key `status`, which can have either the value `warning` or `failed`.
+
+```yaml
+rollback:
+  triggers:
+  - hardening.deployment.finished:
+      selector:
+        matchLabels:
+          status: failed
+  - hardening.release.finished:
+      selector:
+        matchExpressions:
+          - {key: status, operator: In, values: [warning, failed]}
+  tasks:
+  - name: rollback
+```
+
+### Shipyard controller
+
+This KEP also proposes the implementation of a Keptn core component - the so-called *shipyard controller*. This controller implements the functionality described above. Besides, this controller is responsible for keeping sequences atomic. This means that a sequence is executed based on the shipyard, which was configured when the sequence was triggered. In other words, an update of a shipyard (and its containing sequences) does NOT impact ongoing sequences.
+
+### Keptn CloudEvents
+
+Based on the propsed concepts, changes on the Keptn CloudEvents are necessary.
+
+**(1) Reserved space in CloudEvent payload:**
+
+Before the shipyard controller sends out a `[task].triggered` event for a specific task, the controller adds a reserved space to the event payload. This reserved space is named after the task and can be consumed by the unit (Keptn-service) that executes the task. For example, a *deployment* task gets the reserved space of *deployment* in the payload (i.e., data field) of the CloudEvent. Per default, this reserved space is empty. Expect for tasks that have properties defined in the shipyard (e.g., `strategy: blue_green`) - for those tasks the properties are added to the reserved space:
```

Comment:
Typo? Expect for tasks  -> Except for tasks

Reply:
Is a typo. ;) Fixed it.